### ISSUE TITLE
Type inference failure with `std::borrow::Borrow` leads to false positive errors

### ISSUE THREAD

**Issue by cyqsimon** [rust-lang/rust-analyzer#13761]
It's quite difficult to clearly explain what's wrong, so here's a MRE:
```rust
#![allow(dead_code, unused_variables)]
use std::borrow::Borrow;

fn main() {}

struct Foo {}
fn accepts_ref_foo(ref_foo: &Foo) {
    // ...
}

/// This is the MRE.
fn this_errors(foos: impl Borrow<Vec<Foo>>) {
    let vec_of_ref_foo = foos.borrow().iter().collect::<Vec<_>>();

    // 1. observe that RA fails to infer the type of `vec_of_ref_foo` (should be `Vec<&Foo>`)
    // 2. now uncomment the following for loop, and observe that RA thinks `ref_foo` is `Foo` (should be `&Foo`)
    // 3. with the following for loop uncommented, run `cargo check` and `cargo build`,
    //      verify that the compiler sees no problems here

    // for ref_foo in vec_of_ref_foo {
    //     // this destructing statement is critical
    //     // without it, RA sees no issues
    //     let Foo { .. } = ref_foo;
    //     // RA thinks there's an error here where there is none
    //     accepts_ref_foo(ref_foo);
    // }
}

/// Only `Borrow<Vec<Foo>>` is affected, not `Vec<Borrow<Foo>>`.
fn this_is_fine_0(foos: Vec<impl Borrow<Foo>>) {
    let vec_of_ref_foo = foos.iter().map(|foo| foo.borrow()).collect::<Vec<_>>();
    for ref_foo in vec_of_ref_foo {
        let Foo { .. } = ref_foo;
        accepts_ref_foo(ref_foo);
    }
}
```

---

**rust-analyzer version**: `0.3.1309-standalone`

**rustc version**: `rustc 1.65.0 (897e37553 2022-11-02)`

**Comment by cyqsimon** [comment]
![VSCode error screenshot](https://user-images.githubusercontent.com/28627918/207039619-5206885b-4616-4e82-ade9-da21dcdb61b5.png)

**Comment by cyqsimon** [comment]
![VSCode error screenshot](https://user-images.githubusercontent.com/28627918/207040883-fd09ad53-e037-4680-b340-7f46ae2178fe.png)

**Comment by lowr** [comment]
I think this is effectively duplicate of #5514. We cannot resolve `vec_of_ref_foo`'s type correctly as you point out because of that.

rust-analyzer doesn't show type mismatch errors when it comes from unresolved types (i.e. `{unknown}`, but when it sees that `let Foo { .. } = ref_foo` statement there, it coerces the unresolved type to `Foo` because that's the only information it has.

**Comment by cyqsimon** [comment]
I think I'll leave this issue open for now, because it showcases a situation where this limitation leads to an actual error rather than just an annoyance.

**Comment by ChayimFriedman2** [comment]
https://github.com/rust-lang/rust-analyzer/issues/5514 is fixed (new solver, yay!) and this is also gone.

(In fact, we'd had an instance of this in the r-a codebase!)
